### PR TITLE
[Clang][Sema] Differentiate between partial/explicit specializations when diagnosing unexpanded packs

### PR DESCRIPTION
**Pull request by sdkrystian**
This adds `UnexpandedParameterPackContext::UPPC_ExplicitSpecialization` and passes it to `DiagnoseUnexpandedParameterPack` when checking class/variable template explicit specializations. 

Right now we don't check for unexpanded packs in function template explicit specializations at all, so I will be addressing that in a followup PR.

### CODE REVIEW

**Comment by sdkrystian** [comment]
Ping @erichkeane

File: clang/test/CXX/temp/temp.decls/temp.variadic/p5.cpp
Comment (erichkeane) [inline clang/test/CXX/temp/temp.decls/temp.variadic/p5.cpp:373]: This test is not showing any change in behavior? All of these diagnostics are existing behavior (https://godbolt.org/z/xr49ac7Ed). What actual changes does this patch make?

Comment (sdkrystian) [reply clang/test/CXX/temp/temp.decls/temp.variadic/p5.cpp:373]: @erichkeane The existing behavior is to call explicit specializations "partial specializations" in the diagnostic for unexpanded parameter packs. This patch fixes that, so a distinction is made between partial/explicit specializations. For example, the test on line 391 displays different behavior:
```cpp
struct InnerClass<Ts>; // expected-error{{explicit specialization contains unexpanded parameter pack 'Ts'}}
```
Note "explicit specialization" in the diagnostic.

**Review by erichkeane — APPROVED** [review]

**Comment by sdkrystian** [comment]
@erichkeane Unless you would like someone else to review this, could you please merge this when you have time? I don't have write access 🙂

**Pull request merged**
